File: src/cli/config/set.ts
import type { Command } from "commander";
import { printJson, printError } from "../../lib/output.ts";
import { getSettings, updateSettings } from "../../lib/config.ts";
import type { Settings } from "../../lib/config.ts";
import { SETTING_DEFS, VALID_KEYS } from "./index.ts";

export function registerSet(config: Command): void {
  config
    .command("set")
    .argument("<key>", "Setting key")
    .argument("<value>", "Setting value")
    .description("Update a setting")
    .action((key: string, value: string) => {
      const def = SETTING_DEFS[key];
      if (!def) {
        printError(`Unknown setting: ${key}. Valid keys: ${VALID_KEYS.join(", ")}`);
        return;
      }

      let parsed: unknown;
      try {
        parsed = def.parse(value);
      } catch (err) {
        printError((err as Error).message);
        return;
      }

      const [section, field] = def.path;
      const settings = getSettings();
      const sectionObj = (settings[section] ?? {}) as Record<string, unknown>;
      sectionObj[field] = parsed;
      updateSettings({ [section]: sectionObj } as Settings);
      printJson({ [key]: parsed });
    });
}

File: internal/cli/configcmd/set.go
package configcmd

import (
	"fmt"

	"github.com/spf13/cobra"

	"github.com/shhac/lin/internal/config"
	"github.com/shhac/lin/internal/output"
)

func registerSet(cfg *cobra.Command) {
	cmd := &cobra.Command{
		Use:   "set <key> <value>",
		Short: "Update a setting",
		Args:  cobra.ExactArgs(2),
		Run: func(_ *cobra.Command, args []string) {
			key, value := args[0], args[1]
			def, ok := settingDefs[key]
			if !ok {
				output.PrintError(fmt.Sprintf("Unknown setting: %s. Valid keys: %s", key, validKeysStr()))
				return
			}

			parsed, err := def.parse(value)
			if err != nil {
				output.PrintError(err.Error())
				return
			}

			intVal := parsed.(int)
			settings := config.GetSettings()
			def.set(settings, intVal)

			if err := config.UpdateSettings(settings); err != nil {
				output.PrintError(err.Error())
				return
			}
			output.PrintJSON(map[string]any{key: parsed})
		},
	}
	cfg.AddCommand(cmd)
}
